### PR TITLE
fix(PFD): QNH flashing fix

### LINKED ISSUE #3992: QNH blinking even below TRL
**Mod Version**

Dev 2021-03-15T102212

**Describe the bug**

QNH blinking seems linked with approach phase and not the specified TRL as before. For example In USA, TRL is egal to FL180 . So when I switch baro to QNH mode at TRL, I have QNH blinking until I activate Appr phase.

**To Reproduce**

1.  In USA or Japan for example, I specifiy TRL at FL180 
2. During descent when aircraft altitude is below TRL, I switch baro to QNH value but I don't active APPR phase because I'm still far from the destination airport
3. QNH value is blinking in PFD until APPR phase is activated

**Expected behavior**

Like previous mod versions, QNH blinking when baro remains on STD position and when altitude is below specified TRL

**Actual behavior**

QNH blinking when baro is on QNH position and when altitude is below specified TRL

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #3992 

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Vocabulary:
TA: Transition altitude

During Preflight, Takeoff, Climb, Cruise the TA entered on the takeoff performance page is used as source.
For the other flight phases the TA from approach performance page is used.

- When on ground, ensure neither STD or QNH/QFE are blinking.
- In air, ensure
   - QNH/QFE blink when above TA
   - STD blink when below TA

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 1]
```diff
@@ -546,37 +546,35 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
 
-        if ((SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") > 1) && (SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") <= 4)) {
-            if ((transTakeoffAlt <= indicatedAltitude) && (baroMode != "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkQNH();
-            }
-            if ((transTakeoffAlt > indicatedAltitude) && (baroMode == "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkSTD();
-            }
+        // Default TA is 18,000. Will removed with other PR
+        if (SimVar.GetSimVarValue("L:AIRLINER_TRANS_ALT", "Number") == 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:554]: We should use strict compare.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 2]
```diff
@@ -546,37 +546,35 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
+            SimVar.SetSimVarValue("L:AIRLINER_TRANS_ALT", "Number", 18000);
+        }
+        if (SimVar.GetSimVarValue("L:AIRLINER_APPR_TRANS_ALT", "Number") == 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:557]: We should use strict compare.

**Review by MisterChocker — CHANGES_REQUESTED** [review]

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 3]
```diff
@@ -546,37 +546,35 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
+            SimVar.SetSimVarValue("L:AIRLINER_TRANS_ALT", "Number", 18000);
+        }
+        if (SimVar.GetSimVarValue("L:AIRLINER_APPR_TRANS_ALT", "Number") === 0) {
+            SimVar.SetSimVarValue("L:AIRLINER_APPR_TRANS_ALT", "Number", 18000);
         }
 
-        if ((SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") > 4) && (SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") < 7)) {
-            if ((transApprAlt >= indicatedAltitude) && (baroMode == "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkSTD();
+        if (transTakeoffAlt !== 0 && !Simplane.getIsGrounded()) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:561]: I don't see a reason to check in both cases if we are grounded or not.

Also why are we using actual integers for the flight phases?
We have a constant with all flight phases defined.

It could be refactored to:
```js
if (!Simplane.getIsGrounded()) {
   if (transTakeOffAlt !== 0 && phase >= 1 && phase <= 3) {
      this._blinkQNH():
   } else if (transApprAlt !== 0) {
      if (phase === 4 || phase === 5) {
         this._blinkSTD();
      } else if (phase === 6) {
         this._blinkQNH():
      }
   }
}
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 4]
```diff
@@ -546,37 +546,35 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
 
-        if ((SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") > 1) && (SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") <= 4)) {
-            if ((transTakeoffAlt <= indicatedAltitude) && (baroMode != "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkQNH();
-            }
-            if ((transTakeoffAlt > indicatedAltitude) && (baroMode == "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkSTD();
-            }
+        // Default TA is 18,000. Will removed with other PR
+        if (SimVar.GetSimVarValue("L:AIRLINER_TRANS_ALT", "Number") === 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:554]: No need to access the same simvar two rows later again. Just use the const.

```suggestion
        if (transTakeoffAlt === 0) {
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 5]
```diff
@@ -546,37 +546,35 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
+            SimVar.SetSimVarValue("L:AIRLINER_TRANS_ALT", "Number", 18000);
+        }
+        if (SimVar.GetSimVarValue("L:AIRLINER_APPR_TRANS_ALT", "Number") === 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:557]: Same here

```suggestion
        if (transApprAlt === 0) {
```

**Review by MisterChocker — CHANGES_REQUESTED** [review]

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 6]
```diff
@@ -546,34 +546,33 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
-        if (transApprAlt !== 0 && !Simplane.getIsGrounded()) {
-            if (phase === 4 || phase === 5) {
-                if ((transApprAlt >= indicatedAltitude) && (baroMode === "STD")) {
-                    this._blinkSTD();
-                }
-            }
-            if (phase === 6) {
-                if (transApprAlt <= indicatedAltitude && baroMode !== "STD") {
-                    this._blinkQNH();
+            if (destinationTA !== 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:567]: can this be else if?

Comment (Kimbyeoungjang) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:567]: yes!

**Review by MisterChocker — APPROVED** [review]
LGTM

**Comment by ChruutvoLuzi** [comment]
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 8.4.2021
Version of the sim: 1.14.6.0
PR Tested: #4286 
Tier of testing: 2
Changes to observe: QNH Flashing Fix


Testing technique: TryOut
Steps you've made to test the Pull Request:
1.  Spawned C&D at LSZH
2.  Entered Data for a short Testflight (LSZH - LSZH, TA 7500ft, TRL FL070, DEGES2W 28 TRA1A 14, CRZ FL100)
3.  Took off
4. Checked QNH Flashing at TA
5. Descended through TRL and checked QNH again
6. Activated TOGA at 6000ft to trigger GA Phase
7. descended to 6000ft again from FL100

Flight notes (When done a full flight):

Media:
Flashing at TA
https://user-images.githubusercontent.com/70195216/114074458-ad086880-98a4-11eb-8169-06a20330dea6.mp4

Issues (If Any): 
- On Go Around the QNH would blink upon the TRL instead of TA
- When Descending again from a Go Around above the TA/TRL, the QNH does not blink at TRL when passing it

Steps to achieve the issue
go Around and climb above TA
descent back below the TRL

Overall Rating: To be improved, if unrealistic
Conclusions of the testing: Normal Flight Phases work so far, but after a Go Around it seems a bit mixed up.  Cant say if its correct or not, since i am not an A320 Pilot

**Comment by Kimbyeoungjang** [comment]
@ChruutvoLuzi 
Are you insert TRL value to approach phase? that is a TA value which is Honeywell reference.

After go around, climb more than destination TA, will blink(so I think after go around flashing is expect behavior)

and second issue, idk the reason. it dependent by fmgc flight phase. I'll test that part

**Comment by AirbusA320Pro** [comment]
@ChruutvoLuzi Your flight plan test details may be the cause of the confusion. You stated "TA 7500ft, TRL FL070", but you can't have a Transition Level lower than the Transition Altitude. If the TA is 7500ft, (depending on region) the TL could be FL075 or FL080 depending on local QNH.
Also, activating TOGA at 6000ft may not have sequenced the FMGC properly if the aircraft wasn't configured to at least Flap 1. In real life, you wouldn't be using TOGA at that altitude, it would be a "cancelled approach" rather than a Go-Around (obviously unless it's a very high airport elevation. LSZH is only 1400ft, so 6000ft is still 4600ft above the ground).

**Comment by ChruutvoLuzi** [comment]
@AirbusA320Pro yeah looks like i mixed that up, oops. I was at a QNH of 1010 HPa in LSZH which should be TA7000ft and TRL75. I will test it again today then with the correct values

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 7]
```diff
@@ -546,38 +549,41 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
-        }
+        const originTA = SimVar.GetSimVarValue("L:AIRLINER_TRANS_ALT", "Number");
+        const destinationTA = SimVar.GetSimVarValue("L:AIRLINER_APPR_TRANS_ALT", "Number");
+        const phase = SimVar.GetSimVarValue("L:A32NX_FMGC_FLIGHT_PHASE", "Enum");
 
-        if ((SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") > 4) && (SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") < 7)) {
-            if ((transApprAlt >= indicatedAltitude) && (baroMode == "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkSTD();
-            }
+        if (phase === FmgcFlightPhases.GOAROUND) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:556]: I don't understand why we want to do that

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 8]
```diff
@@ -17,6 +17,9 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
         this.hudAPAltitude = 0;
         this.isHud = false;
         this._aircraft = Aircraft.A320_NEO;
+        this.goAround = false;
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:20]: why do we need a local variable for that?

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 9]
```diff
@@ -546,38 +549,41 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
         }
 
-        if (SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Number") == 7) {
-            if ((transApprAlt <= indicatedAltitude) && (baroMode != "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkQNH();
-            }
-            if ((transApprAlt > indicatedAltitude) && (baroMode == "STD") && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") > 3) && (SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "number") < 9)) {
-                this._blinkSTD();
+        if (!Simplane.getIsGrounded()) {
+            if (!this.goAround) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:561]: this accomplishes the same thing and further down change it to else

```suggestion
            if (phase !== FmgcFlightPhases.GOAROUND) {
```

your code also has one flaw, line 572 will never get reached due to the fact that during a goarund we are in the else condition.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 10]
```diff
@@ -546,38 +549,41 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
...
+                        if (destinationTA >= indicatedAltitude && baroMode === "STD") {
+                            this._blinkSTD();
+                        }
+                    } else if (phase === FmgcFlightPhases.GOAROUND) {
+                        if (destinationTA <= indicatedAltitude && baroMode !== "STD") {
+                            this._blinkQNH();
+                        }
+                    }
+                }
+            } else if (this.goAround && destinationTA !== 0) {
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:577]: now this then needs to be an else only.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js [hunk 11]
```diff
@@ -17,6 +17,9 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
         this.hudAPAltitude = 0;
         this.isHud = false;
         this._aircraft = Aircraft.A320_NEO;
+        this.goAround = false;
+        this.currentOrigin = "";
+        this.currentDestination = "";
```
Comment (MisterChocker) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/PFD/AltimeterIndicator.js:22]: for what are these new variables?

**Review by MisterChocker — COMMENTED** [review]
I find it very hard to follow and understand the code, because there also is a logic block that will never get executed.

**Comment by ChruutvoLuzi** [comment]
updated my test report with the new results

**Comment by davidwalschots** [comment]
@Kimbyeoungjang I have created a PR on your repository. Took your logic and improved readability a bit. @MisterChocker helped me with verifying some of the logic and simplification as well.

**Comment by Kimbyeoungjang** [comment]
@davidwalschots 
thank you! I tested that and working beautiful.

**Review by MisterChocker — APPROVED** [review]
LGTM

**Comment by ExampleWasTaken** [comment]
Category: QA Trial
Name: `ExampleWasTaken#0886`
Date of testing: 10/04/2021
Version of the sim: 1.14.6.0
PR Tested: #4286
Tier of testing: 2
Changes to observe:
QNH flashing fix

Testing technique:
1. Loaded on RWY
2. Inserted TO date but left TA untouched
3. Took off climbed to 11000 ft
4. Inserted APPR data and set TA to 7000
5. Performed GA 
6. Landed 

Issues (If Any):
1. QNH did not flash upon reaching automatically set TA
2. QNH did not flash when changing from STD to QNH while being above the set TA

Steps to achieve the issue:
Issue 1:
1. Load on RWY
2. Insert necessary TO data but leave TA untouched
3. Takeoff observe QNH not flashing upon reaching set TA

Issue 2:
1. Fly above set TA with QNH set to STD
2. Change from STD to QNH and observe QNH not flashing.

Overall Rating: To be improved
Conclusions of the testing: Descent works as intended but the takeoff needs some improvement.

**Comment by MisterChocker** [comment]
Fixed the reported issue, where the SimVar used on the PFD is set to 0ft, however the mcdu defaults to 10000ft.

**Comment by St54Kevin** [comment]
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 11/04/2021
Version of the sim: 1.14.6.0
PR Tested: #4286 
Tier of testing: 1
Changes to observe: Baro flashing

**Steps you've made to test the Pull Request:**
1. Loaded up at EDDH and setup on the RW for a full EDDH-EDDH flight, CRZ FL110
2. Entered TA 5000ft on PERF page
3. Took off - STD flashing at 5000ft
4. Set STD, clb to FL110 - Change to QNH will start the flashing
5. TA on APPR page set to 6000ft
6. Early descent to 4000ft (no APPR phase) - QNH flashing at 6000ft
7. Go around - STD flashing at 6000ft

**Media:**

**Issues (If Any):** - 

**Conclusions of the testing:** LGTM

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 11/4/21
Version of the sim: 1.14.6.0
PR Tested: #4286 
Tier of testing: 2
Changes to observe:
QNH flashing

Testing technique: Completed a full flight
Steps you've made to test the Pull Request:
1.  Loaded in 
2. set transition as low as possible (1000) no flashing [good]
3. set transition to 6000 and flew to 6000, flashing seen at 6000
4. set approach page TA to 24000 (cruise) when descend was started the qnh flashed
5. set approach page TA to 6000, flashing occurred at 6000

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.

Issues (If Any): When descending and passing 6000 (TA) I set the QNH to 1023 which was the actual qnh but it continued flashing (STD was flashing as I had just gone below the TA), then moved qnh up by one and then down again by one and flashing stopped. Possibly caused by Departure and arrival qhn being the same so I didn't have to change it.
Steps to achieve the issue: Take off with a certain qnh and then on descend set the same qnh on approach page and set the qnh from standard to the selected qnh (check STD is flashing first) then when switching it will continue flashing until qnh selection is changed

Overall Rating: To be improved
Conclusions of the testing: Small issue found but otherwise worked

**Comment by TLSY2** [comment]
Is this fixed and added to experimental

**Comment by Dreamliner07** [comment]
> Hi, Is this fixed and added to experimental?

Tested yesterday with experimental version and still doesn't work for me.

**Comment by MisterChocker** [comment]
> Hi, Is this fixed and added to experimental?

No it's not.

> Tested yesterday with experimental version and still doesn't work for me.

This is not a fix for experimental

**Comment by TLSY2** [comment]
> > Hi, Is this fixed and added to experimental?
> 
> 
> 
> No it's not.
> 
> 
> 
> > Tested yesterday with experimental version and still doesn't work for me.
> 
> 
> 
> This is not a fix for experimental

What stage is it in? Being worked on or is it done but not for experimental?

**Comment by MisterChocker** [comment]
Just read the above QA reports, then you know the issue.
Again, this is not for experimental.

**Comment by TLSY2** [comment]
> Just read the above QA reports, then you know the issue.
> Again, this is not for experimental.
> This pr is a draft, which means it's being worked on.
> These are all things to be discussed on our discord and not on a pr itself.



> Just read the above QA reports, then you know the issue.
> Again, this is not for experimental.
> This pr is a draft, which means it's being worked on.
> These are all things to be discussed on our discord and not on a pr itself.

I understand thanks!

Im sorry for bothering, I am new to github and dont quite understand everything.

**Comment by MisterChocker** [comment]
Waiting for experimental to land on master.

**Comment by lukecologne** [comment]
Closing, no longer applies as PFD has been redone.
